cmd/kar-controllers/app/server.go

```diff
@@ -62,29 +68,51 @@ func Run(opt *options.ServerOption) error {
 		AgentConfigs: strings.Split(opt.AgentConfigs, ","),
 	}
 
-	jobctrl := queuejob.NewJobController(restConfig, mcadConfig, extConfig)
-	if jobctrl == nil {
-		return nil
-	}
-	jobctrl.Run(neverStop)
 
-	// This call is blocking (unless an error occurs) which equates to <-neverStop
-	err = listenHealthProbe(opt)
+	g, gCtx := errgroup.WithContext(ctx)
+
+	// metrics server
+	metricsServer, err := NewServer(opt.MetricsListenPort, "/metrics", metrics.PrometheusHandler())
 	if err != nil {
 		return err
 	}
 
-	return nil
-}
-
-// Starts the health probe listener
-func listenHealthProbe(opt *options.ServerOption) error {
-	handler := http.NewServeMux()
-	handler.Handle("/healthz", &health.Handler{})
-	err := http.ListenAndServe(opt.HealthProbeListenAddr, handler)
+	healthServer, err := NewServer(opt.HealthProbeListenPort, "/healthz", healthHandler())
 	if err != nil {
 		return err
 	}
 
-	return nil
+	jobctrl := queuejob.NewJobController(restConfig, mcadConfig, extConfig)
+	if jobctrl == nil {
+		return fmt.Errorf("failed to create a job controller")
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	g.Go(func() error {
+		defer wg.Done()
+		jobctrl.Run(gCtx.Done())
```

Comment:
@dimakis,
The `jobctrl.Run` function is asysnchrouous, it will fire off a few `deamon` goroutines and return, thusly triggering the defer call and efectively shutting down the process. (the container is constant crashbackoff loop). I would suggest making the jobctrl.Run be synchrounous or revisiting this implementation.

cc: @astefanutti

Reply:
This is what I see when running the controller:

![Screenshot 2023-09-19 at 09 52 46](https://github.com/project-codeflare/multi-cluster-app-dispatcher/assets/109964136/fc94b5fa-6ced-42c2-a4b4-3f3a28e4577e)

Reply:
@z103cb thanks, very good catch. That Go routine should block for the group context channel to close.